src/main/java/org/candlepin/subscriptions/metering/service/prometheus/PrometheusMetricsProperties.java

```diff
@@ -35,50 +42,80 @@
+    tagProfile.measurementsByTag(productTag).forEach(metric -> metrics.put(metric, openshift));
+    return metrics;
   }
 
-  // ENT-3835 will change the data structures used here and should refactor this method as needed
-  public String getEnabledAccountPromQLforProductProfile(String productProfileId) {
-    // NOTE(khowell): doesn't make sense for a given product profile (e.g. OSD) to have different
-    // queries per-metric. Grabbing the first non-empty one for now.
-    return getSupportedMetricsForProduct(productProfileId).values().stream()
-        .map(MetricProperties::getEnabledAccountPromQL)
-        .filter(StringUtils::hasText)
-        .findFirst()
-        .orElseThrow();
+  public Collection<String> getMetricsEnabledProductTags() {
+    return tagProfile.getTagsWithPrometheusEnabledLookup();
   }
 
-  // ENT-3835 will change the data structures used here and should refactor this method as needed
-  public Integer getMetricsTimeoutForProductProfile(String productProfileId) {
-    // NOTE(khowell): doesn't make sense for a given product profile (e.g. OSD) to have different
+  public Integer getMetricsTimeoutForProductTag(String productTag) {
+    // NOTE(khowell): doesn't make sense for a given product tag (e.g. OSD) to have different
     // metrics timeouts. Grabbing the first one for now.
-    return getSupportedMetricsForProduct(productProfileId).values().stream()
+    return getSupportedMetricsForProduct(productTag).values().stream()
         .map(MetricProperties::getQueryTimeout)
         .findFirst()
         .orElseThrow();
   }
 
-  // ENT-3835 will change the data structures used here and should refactor this method as needed
-  public Integer getRangeInMinutesForProductProfile(String productProfileId) {
-    return getSupportedMetricsForProduct(productProfileId).values().stream()
+  public Integer getRangeInMinutesForProductTag(String productTag) {
+    return getSupportedMetricsForProduct(productTag).values().stream()
         .map(MetricProperties::getRangeInMinutes)
         .findFirst()
         .orElseThrow();
   }
+
+  public Optional<String> getQueryTemplate(String templateKey) {
+    return queryTemplates.containsKey(templateKey)
+        ? Optional.of(queryTemplates.get(templateKey))
+        : Optional.empty();
+  }
+
+  public Optional<String> getAccountQueryTemplate(String templateKey) {
+    return accountQueryTemplates.containsKey(templateKey)
+        ? Optional.of(accountQueryTemplates.get(templateKey))
+        : Optional.empty();
+  }
+
+  public Optional<TagMetric> getTagMetric(String productTag, Uom metric) {
+    if (!StringUtils.hasText(productTag) || Objects.isNull(metric)) {
+      return Optional.empty();
+    }
+
+    return tagProfile.getTagMetrics().stream()
+        .filter(x -> productTag.equals(x.getTag()) && metric.equals(x.getUom()))
+        .findFirst();
+  }
+
+  public Optional<TagMetaData> getTagMetadata(String productTag) {
```

Comment:
I'd rather see this method in TagProfile, along with precomputation (produce view in constructor).

Reply:
DONE